pennylane/compiler/python_compiler/transforms/mbqc/outline_state_evolution.py

```diff
@@ -0,0 +1,465 @@
+
+        # collect required inputs for the state evolution funcOp
+        required_inputs = self._collect_required_inputs_for_state_evolution_func(ops_to_clone)
+
+        # colloect required outputs for the state evolution funcOp
+        required_outputs = self._collect_required_outputs(ops_to_clone, terminal_boundary_op)
+
+        register_inputs = []
+        other_inputs = []
+        for val in required_inputs:
+            if isinstance(val.type, quantum.QuregType):
+                register_inputs.append(val)
+            else:
+                other_inputs.append(val)
+
+        register_outputs = []
+        other_outputs = []
+        for val in required_outputs:
+            if isinstance(val.type, quantum.QuregType):
+                register_outputs.append(val)
+            else:
+                other_outputs.append(val)
+
+        ordered_inputs = register_inputs + other_inputs
+        ordered_outputs = register_outputs + other_outputs
+
+        input_types = [val.type for val in ordered_inputs]
+        output_types = [val.type for val in ordered_outputs]
+        fun_type = builtin.FunctionType.from_lists(input_types, output_types)
+
+        # create a new func.func op and insert it into the IR
+        state_evolution_func = func.FuncOp(
+            self.original_func_op.sym_name.data + ".state_evolution", fun_type, visibility="public"
+        )
+        rewriter.insert_op(state_evolution_func, InsertPoint.at_end(self.module.body.block))
+
+        # TODOs: how to define the `value_mapper` arg is not stated in the xdl.core module [here](https://github.com/xdslproject/xdsl/blob/e1301e0204bcf6ea5ed433e7da00bee57d07e695/xdsl/ir/core.py#L1429)_.
+        # It looks like storing ssa value to be cloned would maintain the dependency relationship required to build the new DAG for the new ops.
+        block = state_evolution_func.regions[0].block
+        value_mapper = {}
+        for input, block_arg in zip(ordered_inputs, block.args):
+            value_mapper[input] = block_arg
+
+        self._clone_operations_to_block(ops_to_clone, block, value_mapper)
+        self._add_return_statement(block, ordered_outputs, value_mapper)
+
+        self.required_inputs = ordered_inputs
+        self.required_outputs = ordered_outputs
+        self.alloc_op = alloc_op
+        self.terminal_boundary_op = terminal_boundary_op
+        self.state_evolution_func = state_evolution_func
+
+    def _find_evolution_range(self):
+        """find alloc_op and terminal_boundary_op"""
+        alloc_op = None
+        terminal_boundary_op = None
+
+        for op in self.original_func_op.body.walk():
+            if isinstance(op, quantum.AllocOp):
+                alloc_op = op
+            elif hasattr(op, "attributes") and "terminal_boundary" in op.attributes:
+                terminal_boundary_op = op
+
+        if not alloc_op or not terminal_boundary_op:
```

Comment:
```suggestion
        if not (alloc_op and terminal_boundary_op):
```
Should we combine these? Or would the short-circuit dual condition be better?